### PR TITLE
AI storage components

### PR DESCRIPTION
**Pull request by ghost**
Potato AI, anyone?

:cl: More Robust Than You
refactor: Intellicards now use components
/:cl:

### CODE REVIEW

**Comment by MoreRobustThanYou** [comment]
Pinging @Cyberboss for review since components are kinda his thing

File: code/datums/components/ai_storage.dm [hunk 1]
```diff
@@ -0,0 +1,28 @@
+/datum/component/ai_storage
+	var/mob/living/silicon/ai/AI
+	var/wiping = FALSE
+
+/datum/component/ai_storage/Initialize(_wiping=FALSE)
+	wiping = _wiping
+	RegisterSignal(COMSIG_ITEM_ATTACK_OBJ, .proc/attack_obj)
+
+/datum/component/ai_storage/proc/wipe()
+	INVOKE_ASYNC(src, .proc/wipe_loop)
```
Comment (Xhuis) [inline code/datums/components/ai_storage.dm]: There's no reason not to just have the contents of `wipe_loop()` in here, isn't there? This is literally a proc whose purpose is to call another proc!

Comment (MoreRobustThanYou) [reply code/datums/components/ai_storage.dm]: It's purpose is to INVOKE_ASYNC the proc, as INVOKE_ASYNC allows the current proc to continue.

File: code/datums/components/ai_storage.dm [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+/datum/component/ai_storage/proc/wipe()
+	INVOKE_ASYNC(src, .proc/wipe_loop)
+
+/datum/component/ai_storage/proc/wipe_loop()
+	if(AI && AI.loc == src)
+		to_chat(AI, "<span class='userdanger'>Your core files are being wiped!</span>")
+		while(AI.stat != DEAD && wiping)
+			AI.adjustOxyLoss(1)
+			AI.updatehealth()
+			sleep(5)
```
Comment (Cyberboss) [inline code/datums/components/ai_storage.dm]: I'm not letting you get away with this. Do it properly with processing

File: code/datums/components/ai_storage.dm [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+		wiping = FALSE
+
+/datum/component/ai_storage/proc/attack_obj(obj/target, mob/living/user)
+	if(!user.Adjacent(target) || !target )
+		return
```
Comment (Cyberboss) [inline code/datums/components/ai_storage.dm]: completely uneeded checks

File: code/datums/components/ai_storage.dm [hunk 4]
```diff
@@ -0,0 +1,28 @@
+/datum/component/ai_storage
+	var/mob/living/silicon/ai/AI
```
Comment (Cyberboss) [inline code/datums/components/ai_storage.dm]: Should be a weakref

File: code/game/gamemodes/objective_items.dm [hunk 5]
```diff
@@ -130,11 +130,12 @@
...
 
-/datum/objective_item/steal/functionalai/check_special_completion(obj/item/device/aicard/C)
-	for(var/mob/living/silicon/ai/A in C)
-		if(isAI(A) && A.stat != DEAD) //See if any AI's are alive inside that card.
-			return 1
-	return 0
+/datum/objective_item/steal/functionalai/check_special_completion(obj/C)
+	if(istype(C, /obj/item/device/aicard))
+		for(var/mob/living/silicon/ai/A in C)
+			if(isAI(A) && A.stat != DEAD) //See if any AI's are alive inside that card.
```
Comment (Cyberboss) [inline code/game/gamemodes/objective_items.dm]: remove the isAI check

File: code/datums/components/ai_storage.dm (hunk 3 above)
Comment (Cyberboss) [reply code/datums/components/ai_storage.dm]: Also where is the check if it's attacking an AI core?

File: code/game/machinery/computer/aifixer.dm [hunk 6]
```diff
@@ -115,10 +115,14 @@
 		else
 			add_overlay("ai-fixer-empty")
 
-/obj/machinery/computer/aifixer/transfer_ai(interaction, mob/user, mob/living/silicon/ai/AI, obj/item/device/aicard/card)
+/obj/machinery/computer/aifixer/transfer_ai(interaction, mob/user, mob/living/silicon/ai/AI, obj/card)
 	if(!..())
 		return
 	//Downloading AI from card to terminal.
+	var/datum/component/ai_storage/C = card.GetComponent(/datum/component/ai_storage)
```
Comment (Cyberboss) [inline code/game/machinery/computer/aifixer.dm]: GET_COMPONENT_FROM(/datum/component/ai_storage, C, card)

File: code/game/mecha/mecha.dm [hunk 7]
```diff
@@ -661,10 +661,12 @@
 			return
 		to_chat(user, "<a href='?src=[REF(user)];ai_take_control=[REF(src)]'><span class='boldnotice'>Take control of exosuit?</span></a><br>")
 
-/obj/mecha/transfer_ai(interaction, mob/user, mob/living/silicon/ai/AI, obj/item/device/aicard/card)
+/obj/mecha/transfer_ai(interaction, mob/user, mob/living/silicon/ai/AI, obj/card)
 	if(!..())
 		return
-
+	var/datum/component/ai_storage/C = card.GetComponent(/datum/component/ai_storage)
```
Comment (Cyberboss) [inline code/game/mecha/mecha.dm]: again

File: code/game/objects/structures/ai_core.dm [hunk 8]
```diff
@@ -253,28 +253,34 @@ That prevents a few funky behaviors.
 //The type of interaction, the player performing the operation, the AI itself, and the card object, if any.
 
 
-/atom/proc/transfer_ai(interaction, mob/user, mob/living/silicon/ai/AI, obj/item/device/aicard/card)
-	if(istype(card))
-		if(card.flush)
-			to_chat(user, "<span class='boldannounce'>ERROR</span>: AI flush is in progress, cannot execute transfer protocol.")
-			return 0
-	return 1
+/atom/proc/transfer_ai(interaction, mob/user, mob/living/silicon/ai/AI, obj/card)
```
Comment (Cyberboss) [inline code/game/objects/structures/ai_core.dm]: should be a component proc

File: code/game/objects/structures/ai_core.dm [hunk 9]
```diff
@@ -253,28 +253,34 @@ That prevents a few funky behaviors.
...
+	if(!C)
+		return
+	if(C.wiping)
+		to_chat(user, "<span class='boldannounce'>ERROR</span>: AI flush is in progress, cannot execute transfer protocol.")
+		return FALSE
+	return TRUE
 
 
-/obj/structure/AIcore/transfer_ai(interaction, mob/user, mob/living/silicon/ai/AI, obj/item/device/aicard/card)
+/obj/structure/AIcore/transfer_ai(interaction, mob/user, mob/living/silicon/ai/AI, obj/card)
```
Comment (Cyberboss) [inline code/game/objects/structures/ai_core.dm]: The core itself should have a storage component which does the transform with a callback

File: code/modules/mob/living/silicon/ai/ai.dm [hunk 10]
```diff
@@ -774,9 +774,12 @@
 		return
 	set_autosay()
 
-/mob/living/silicon/ai/transfer_ai(interaction, mob/user, mob/living/silicon/ai/AI, obj/item/device/aicard/card)
+/mob/living/silicon/ai/transfer_ai(interaction, mob/user, mob/living/silicon/ai/AI, obj/card)
```
Comment (Cyberboss) [inline code/modules/mob/living/silicon/ai/ai.dm]: ditto here

File: code/modules/power/apc.dm [hunk 11]
```diff
@@ -979,8 +979,11 @@
 				P.switch_mode_to(TRACK_NUKE_DISK) //Pinpointers go back to tracking the nuke disk
 				P.alert = FALSE
 
-/obj/machinery/power/apc/transfer_ai(interaction, mob/user, mob/living/silicon/ai/AI, obj/item/device/aicard/card)
-	if(card.AI)
+/obj/machinery/power/apc/transfer_ai(interaction, mob/user, mob/living/silicon/ai/AI, obj/card)
```
Comment (Cyberboss) [inline code/modules/power/apc.dm]: and again

**Review by Cyberboss — CHANGES_REQUESTED** [review]
Needs an InheritComponent or maybe you should set dupe_mode to ALLOWED

File: code/datums/components/ai_storage.dm (hunk 1 above)
Comment (Cyberboss) [reply code/datums/components/ai_storage.dm]: #33446, change it

**Comment by PKPenguin321** [comment]
Would this be good to go with the changes cyberboss listed or would there be significantly more? AI stuck to a potato battery would be funny

**Comment by Cyberboss** [comment]
Not significantly, after what I said it's pretty good to go

**Pull request closed without merging**